### PR TITLE
[move] Ban extraneous module bytes and metadata

### PR DESCRIPTION
**Pull request by tnowacki**
## Description 

- Ban extraneous module bytes
- Ban metadata

## Test Plan 

- New tests 

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: crates/sui-protocol-config/src/lib.rs [hunk 1]
```diff
@@ -1099,6 +1106,7 @@ impl ProtocolConfig {
                 let mut cfg = Self::get_for_version_impl(version - 1);
                 cfg.feature_flags
                     .disallow_change_struct_type_params_on_upgrade = true;
+                cfg.feature_flags.no_extraneous_module_bytes = true;
```
Comment (tnowacki) [inline crates/sui-protocol-config/src/lib.rs:1109]: @tzakian has 8 gone out yet?

File: external-crates/move/move-binary-format/src/deserializer.rs [hunk 2]
```diff
@@ -35,23 +35,34 @@ impl CompiledScript {
...
+            /* check_no_extraneous_bytes */ false,
+        )
     }
 
-    /// Deserialize a &[u8] slice into a `CompiledModule` instance, up to the specified version.
-    pub fn deserialize_with_max_version(
+    /// Deserialize a &[u8] slice into a `CompiledModule` instance with settings
+    /// - Can specify up to the specified version.
+    /// - Can specify if the deserializer should error on trailing bytes
+    pub fn deserialize_with_config(
```
Comment (dariorussi) [inline external-crates/move/move-binary-format/src/deserializer.rs:48]: this is just for another day but I'd think we should rename `deserialize` to `deserialize_with_defaults` and change `deserialize_with_config` to `deserialize`

File: external-crates/move/move-binary-format/src/deserializer.rs [hunk 3]
```diff
@@ -572,7 +590,7 @@ fn build_common_tables(
                 load_constant_pool(binary, table, common.get_constant_pool())?;
             }
             TableType::METADATA => {
-                if binary.version() < VERSION_5 {
+                if binary.check_no_extraneous_bytes() || binary.version() < VERSION_5 {
```
Comment (dariorussi) [inline external-crates/move/move-binary-format/src/deserializer.rs:593]: so this section is banned from now on, right?

**Review by dariorussi — APPROVED** [review]
this looks good to me and I like that we add those checks.
Not sure if others have concerns or opinions.
I think @oxade may know if version 8 is out and when

File: external-crates/move/move-binary-format/src/deserializer.rs (hunk 3 above)
Comment (tnowacki) [reply external-crates/move/move-binary-format/src/deserializer.rs:593]: correct, it will give the error MALFORMED if it sees a TableType::METADATA

**Review by oxade — CHANGES_REQUESTED** [review]
I think PV 8 is already cut.
Change to 9

File: crates/sui-protocol-config/src/lib.rs (hunk 1 above)
Comment (oxade) [reply crates/sui-protocol-config/src/lib.rs:1109]: > 

I think PV 8 is already cut.
Change to 9

**Comment by tnowacki** [comment]
> I think PV 8 is already cut. Change to 9

Yarp, that's why it was set back to draft

**Review by oxade — APPROVED** [review]

**Pull request merged**